assets/extensions/store.js

```diff
@@ -0,0 +1,292 @@
+
+		try {
+			const response = yield apiFetch( {
+				path: '/sensei-internal/v1/sensei-extensions/update',
+				method: 'POST',
+				data: { plugins: slugs },
+			} );
+
+			yield actions.setError( null );
+			yield actions.setExtensions(
+				{
+					extensions: response.completed,
+					wccom_connected: response.wccom_connected,
+				},
+				true
+			);
+
+			yield dispatch( 'core/notices' ).createNotice(
+				'success',
+				__( 'Update completed succesfully!', 'sensei-lms' ),
+				{
+					type: 'snackbar',
+				}
+			);
+		} catch ( error ) {
+			yield actions.setExtensionsStatus( slugs, '' );
+			yield actions.setError(
+				sprintf(
+					// translators: Placeholder is underlying error message.
+					__(
+						'There was an error while updating the plugin: %1$s.',
+						'sensei-lms'
+					),
+					error.message
+				)
+			);
+		} finally {
+			// Update queue extensions, if exists.
+			const queuedExtensions = yield select(
+				EXTENSIONS_STORE
+			).getExtensionsByStatus( STATUS.IN_QUEUE );
+
+			if ( 0 === queuedExtensions.length ) {
+				return;
+			}
+
+			yield actions.updateExtensions( queuedExtensions );
+		}
+	},
+
+	/**
+	 * Set extensions in progress.
+	 *
+	 * @param {string} slugs  Extensions in progress.
+	 * @param {string} status Status.
+	 */
+	setExtensionsStatus( slugs, status ) {
+		return {
+			type: 'SET_EXTENSIONS_STATUS',
+			slugs,
+			status,
+		};
+	},
+
+	/**
+	 * Set the extensions layout.
+	 *
+	 * @param {Object} obj        Layout object.
+	 * @param {Array}  obj.layout Extensions layout.
+	 */
+	setLayout( { layout = [] } ) {
+		return {
+			type: 'SET_LAYOUT',
+			layout,
+		};
+	},
+
+	/**
+	 * Set the error message.
+	 *
+	 * @param {string} error The error.
+	 */
+	setError( error ) {
+		return {
+			type: 'SET_ERROR',
+			error,
+		};
+	},
+};
+
+/**
+ * Extension store selectors.
+ */
+const selectors = {
+	getExtensions: ( { extensions, entities } ) =>
+		extensions.map( ( slug ) => entities.extensions[ slug ] ),
+	getExtensionsByStatus: ( args, status ) =>
+		selectors
+			.getExtensions( args )
+			.filter( ( extension ) => status === extension.status ),
+	getLayout: ( { layout } ) => layout,
+	getError: ( { error } ) => error,
+};
+
+/**
+ * Extension store resolvers.
+ */
+const resolvers = {
+	/**
+	 * Loads the extensions during initialization.
+	 */
+	*getExtensions() {
+		const response = yield apiFetch( {
+			path: '/sensei-internal/v1/sensei-extensions?type=plugin',
+		} );
+
+		return actions.setExtensions( response );
+	},
+
+	/**
+	 * Loads the extensions layout.
+	 */
+	*getLayout() {
+		const response = yield apiFetch( {
+			path: '/sensei-internal/v1/sensei-extensions/layout',
+		} );
+
+		return actions.setLayout( response );
+	},
+};
+
+/**
+ * Extensions store reducer.
+ *
+ * @param {Object} state  The store state.
+ * @param {Object} action The action to handle.
+ */
+const reducer = (
+	state = {
+		extensions: [],
+		entities: { extensions: {} },
```

Comment:
This is pretty confusing & seems to lead to a lot of extra hoops. We could just have an `extensions: { }` object, and have a `getExtensionSlugs` selector mapping that to just the slug list if it's needed anywhere. There isn't any significance to the extension order that would warrant storing them as a list too. (And if there were, we should use more explicit naming to communicate the difference between the list and the entities map)

Reply:
When we filter the extensions by Free, or Installed (the tabs), we follow the original order. It's the only part where it's important, but I think it's still important.

> we should use more explicit naming to communicate the difference between the list and the entities map

I thought about calling the `extensions` array as `list`, but I think it can be confusing. Do you have a name suggestion you think would be more clear?

Reply:
Just adding here the inspiration of this approach: https://github.com/paularmstrong/normalizr

I used to use it in many projects and it uses to work very well. Imagine a case where we want to store some layout entities too, for example. We would have:

```js
{
  extensions: [ 'extension-1', 'extension-2'... ],
  layout: [ 'layout-1', 'layout-2'... ],
  entities: {
    extensions: {
      'extension-1': { title: 'Layout' },
      ...
    }
    layout: {
      'layout-1': { title: 'Layout' },
      ...
    }
  },
}
```

Reply:
Maybe `extensionsOrder` for the list? Or adding a `byId` object to the entities map. The issue is it's hard to distinguish when working with both `extensions` and `entities.extensions` the same time — the names of the properties and the names of the functions manipulating them could help clear things up. Might also be worth to add doc comments to the state object.

If the list is staying, I think `setExtensions` could be broken up, along the `onlyEntities` flag: maybe an `updateExtensionEntities` action when applying the results to some of the extensions?

Reply:
I renamed this to `extensionSlugs`, because `extensionsOrder` I think is more restricted to only order purpose (it's also responsible by show or not the extension, for example).

And also refactored the entities part. See https://github.com/Automattic/sensei/pull/4205/commits/06ce094c3967e55f0d0fb8bacac38ebb545aea35 (there are other parts in other commits too :P)